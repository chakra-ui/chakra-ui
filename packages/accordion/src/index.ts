export * from "./accordion"
export * from "./use-accordion"
