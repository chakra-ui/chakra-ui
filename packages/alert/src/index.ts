export { Alert, type AlertProps } from "./alert"
export { useAlertStyles, type AlertStatus } from "./alert-context"
export {
  AlertDescription,
  type AlertDescriptionProps,
} from "./alert-description"
export { AlertIcon, type AlertIconProps } from "./alert-icon"
export { AlertTitle, type AlertTitleProps } from "./alert-title"
