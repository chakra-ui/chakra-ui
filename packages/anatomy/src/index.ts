import { anatomy } from "@chakra-ui/theme-tools"

/**
 * **Accordion anatomy**
 * - Root: the root container of the accordion
 * - Container: the accordion item contains the button and panel
 * - Button: the button is the trigger for the panel
 * - Panel: the panel is the content of the accordion item
 * - Icon: the expanded/collapsed icon
 */
export const accordionAnatomy = anatomy("accordion")
  .parts("root", "container", "button", "panel")
  .extend("icon")

/**
 * **Alert anatomy**
 * - Title: the alert's title
 * - Description: the alert's description
 * - Icon: the alert's icon
 */
export const alertAnatomy = anatomy("alert")
  .parts("title", "description", "container")
  .extend("icon")

/**
 * **Avatar anatomy**
 * - Container: the container for the avatar
 * - Label: the avatar initials text
 * - Excess Label: the label or text that represents excess avatar count.
 * Typically used in avatar groups.
 * - Group: the container for the avatar group
 */
export const avatarAnatomy = anatomy("avatar")
  .parts("label", "badge", "container")
  .extend("excessLabel", "group")

/**
 * **Breadcrumb anatomy**
 * - Item: the container for a breadcrumb item
 * - Link: the element that represents the breadcrumb link
 * - Container: the container for the breadcrumb items
 * - Separator: the separator between breadcrumb items
 */
export const breadcrumbAnatomy = anatomy("breadcrumb")
  .parts("link", "item", "container")
  .extend("separator")

export const buttonAnatomy = anatomy("button").parts()

export const checkboxAnatomy = anatomy("checkbox")
  .parts("control", "icon", "container")
  .extend("label")

export const circularProgressAnatomy = anatomy("progress")
  .parts("track", "filledTrack")
  .extend("label")

export const drawerAnatomy = anatomy("drawer")
  .parts("overlay", "dialogContainer", "dialog")
  .extend("header", "closeButton", "body", "footer")

export const editableAnatomy = anatomy("editable").parts(
  "preview",
  "input",
  "textarea",
)

export const formAnatomy = anatomy("form").parts(
  "container",
  "requiredIndicator",
  "helperText",
)

export const formErrorAnatomy = anatomy("formError").parts("text", "icon")

export const inputAnatomy = anatomy("input").parts("addon", "field", "element")

export const listAnatomy = anatomy("list").parts("container", "item", "icon")

export const menuAnatomy = anatomy("menu")
  .parts("button", "list", "item")
  .extend("groupTitle", "command", "divider")

export const modalAnatomy = anatomy("modal")
  .parts("overlay", "dialogContainer", "dialog")
  .extend("header", "closeButton", "body", "footer")

export const numberInputAnatomy = anatomy("numberinput").parts(
  "root",
  "field",
  "stepperGroup",
  "stepper",
)

export const pinInputAnatomy = anatomy("pininput").parts("field")

export const popoverAnatomy = anatomy("popover")
  .parts("content", "header", "body", "footer")
  .extend("popper", "arrow", "closeButton")

export const progressAnatomy = anatomy("progress").parts(
  "label",
  "filledTrack",
  "track",
)

export const radioAnatomy = anatomy("radio").parts(
  "container",
  "control",
  "label",
)

export const selectAnatomy = anatomy("select").parts("field", "icon")

export const sliderAnatomy = anatomy("slider").parts(
  "container",
  "track",
  "thumb",
  "filledTrack",
)

export const statAnatomy = anatomy("stat").parts(
  "container",
  "label",
  "helpText",
  "number",
  "icon",
)

export const switchAnatomy = anatomy("switch").parts(
  "container",
  "track",
  "thumb",
)

export const tableAnatomy = anatomy("table").parts(
  "table",
  "thead",
  "tbody",
  "tr",
  "th",
  "td",
  "tfoot",
  "caption",
)

export const tabsAnatomy = anatomy("tabs").parts(
  "root",
  "tab",
  "tablist",
  "tabpanel",
  "tabpanels",
  "indicator",
)

/**
 * **Tag anatomy**
 * - Container: the container for the tag
 * - Label: the text content of the tag
 * - closeButton: the close button for the tag
 */
export const tagAnatomy = anatomy("tag").parts(
  "container",
  "label",
  "closeButton",
)
