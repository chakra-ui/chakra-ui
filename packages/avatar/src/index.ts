export * from "./avatar"
export * from "./avatar-group"
