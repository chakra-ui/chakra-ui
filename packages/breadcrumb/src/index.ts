export { Breadcrumb, type BreadcrumbProps } from "./breadcrumb"
export { BreadcrumbItem, type BreadcrumbItemProps } from "./breadcrumb-item"
export {
  BreadcrumbSeparator,
  type BreadcrumbSeparatorProps,
} from "./breadcrumb-separator"
export { BreadcrumbLink, type BreadcrumbLinkProps } from "./breadcrumb-link"
export { useBreadcrumbStyles } from "./breadcrumb-context"
export type { BreadcrumbOptions } from "./breadcrumb-types"
