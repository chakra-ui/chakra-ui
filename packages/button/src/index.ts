export * from "./button"
export * from "./button-group"
export * from "./icon-button"
export * from "./button-spinner"
