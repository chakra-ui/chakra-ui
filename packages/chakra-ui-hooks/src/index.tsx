export * from "./useAccordion/useAccordion";
export { default as useAriaHidden } from "./useModal/utils/useAriaHidden";
export { default as useBlurOutside } from "./useBlurOutside";

export { default as useCheckbox } from "./useCheckbox/useCheckbox";
export * from "./useCheckboxGroup/useCheckboxGroup";
export {
  default as useCheckboxGroup,
} from "./useCheckboxGroup/useCheckboxGroup";
export { default as useCounter } from "./useCounter/useCounter";
export { default as useClipboard } from "./useClipboard/useClipboard";
export { default as useControllableValue } from "./useControllableValue";
export { default as createCtx } from "./useCreateContext";

export { default as useDisclosure } from "./useDisclosure/useDisclosure";

export { default as useIsomorphicEffect } from "./useIsomorphicEffect";

export { default as useFocusEffect } from "./useFocusEffect/useFocusEffect";
export { default as useFocusOnHide } from "./useFocusOnHide";
export { default as useFocusOnShow } from "./useFocusOnShow";
export * from "./useForkRef";

export { default as useId } from "./useId";
export { default as useIds } from "./useIds";

export {
  default as useLockBodyScroll,
} from "./useLockBodyScroll/useLockBodyScroll";
export { default as useLogger } from "./useLogger";

// export * from "./useNumberInput/useNumberInput";
// export { default as useNumberInput } from "./useNumberInput/useNumberInput";

export * from "./usePopper/usePopper";
export { default as usePopper } from "./usePopper/usePopper";
export { default as usePrevious } from "./usePrevious";

export * from "./useRadioGroup/useRadioGroup";
export { default as useRadioGroup } from "./useRadioGroup/useRadioGroup";
export { default as useRapidKeyDown } from "./useRapidKeydown/useRapidKeyDown";

export * from "./useSelection";
export * from "./useSlider/useSlider";

export * from "./useTabbable";
export * from "./useTabs";
export { default as useTabbable } from "./useTabbable";

export { default as useUpdateEffect } from "./useUpdateEffect";
