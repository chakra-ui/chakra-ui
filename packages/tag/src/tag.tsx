import { Icon, IconProps } from "@chakra-ui/icon"
import {
  chakra,
  forwardRef,
  omitThemingProps,
  StylesProvider,
  SystemStyleObject,
  ThemingProps,
  useMultiStyleConfig,
  useStyles,
  HTMLChakraProps,
} from "@chakra-ui/system"
import { __DEV__ } from "@chakra-ui/utils"
import * as React from "react"

export interface TagProps
  extends HTMLChakraProps<"span">,
    ThemingProps<"Tag"> {}

/**
 * The tag component is used to label or categorize UI elements.
 * To style the tag globally, change the styles in `theme.components.Tag`
 * @see Docs https://chakra-ui.com/tag
 */
export const Tag = forwardRef<TagProps, "span">((props, ref) => {
  const styles = useMultiStyleConfig("Tag", props)
  const ownProps = omitThemingProps(props)

  const containerStyles: SystemStyleObject = {
    display: "inline-flex",
    verticalAlign: "top",
    alignItems: "center",
    maxWidth: "100%",
    ...styles.container,
  }

  return (
    <StylesProvider value={styles}>
      <chakra.span ref={ref} {...ownProps} __css={containerStyles} />
    </StylesProvider>
  )
})

if (__DEV__) {
  Tag.displayName = "Tag"
}

export interface TagLabelProps extends HTMLChakraProps<"span"> {}

export const TagLabel = forwardRef<TagLabelProps, "span">((props, ref) => {
  const styles = useStyles()
  return <chakra.span ref={ref} isTruncated {...props} __css={styles.label} />
})

if (__DEV__) {
  TagLabel.displayName = "TagLabel"
}

export const TagLeftIcon = forwardRef<IconProps, "svg">((props, ref) => (
  <Icon ref={ref} verticalAlign="top" marginEnd="0.5rem" {...props} />
))

if (__DEV__) {
  TagLeftIcon.displayName = "TagLeftIcon"
}

export const TagRightIcon = forwardRef<IconProps, "svg">((props, ref) => (
  <Icon ref={ref} verticalAlign="top" marginStart="0.5rem" {...props} />
))

if (__DEV__) {
  TagRightIcon.displayName = "TagRightIcon"
}

const TagCloseIcon: React.FC<IconProps> = (props) => (
  <Icon verticalAlign="inherit" viewBox="0 0 512 512" {...props}>
    <path
      fill="currentColor"
      d="M289.94 256l95-95A24 24 0 00351 127l-95 95-95-95a24 24 0 00-34 34l95 95-95 95a24 24 0 1034 34l95-95 95 95a24 24 0 0034-34z"
    />
  </Icon>
)

if (__DEV__) {
  TagCloseIcon.displayName = "TagCloseIcon"
}

export interface TagCloseButtonProps
  extends Omit<HTMLChakraProps<"button">, "disabled"> {
  isDisabled?: boolean
}

/**
 * TagCloseButton is used to close "remove" the tag
 * @see Docs https://chakra-ui.com/tag
 */
export const TagCloseButton: React.FC<TagCloseButtonProps> = (props) => {
  const { isDisabled, children, ...rest } = props

  const styles = useStyles()

  const btnStyles: SystemStyleObject = {
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    outline: "0",
    ...styles.closeButton,
  }

  return (
    <chakra.button
      {...rest}
      type="button"
      aria-label="close"
      disabled={isDisabled}
      __css={btnStyles}
    >
      {children || <TagCloseIcon />}
    </chakra.button>
  )
}

if (__DEV__) {
  TagCloseButton.displayName = "TagCloseButton"
}
