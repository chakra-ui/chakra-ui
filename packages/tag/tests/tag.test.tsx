import { render } from "@chakra-ui/test-utils/src"
import * as React from "react"
import { Tag, TagCloseButton, TagLabel, TagLeftIcon } from "../src"

test("Tag (only) renders correctly", () => {
  const { asFragment } = render(<Tag>A</Tag>)
  expect(asFragment()).toMatchSnapshot()
})

test("Tag with close button renders correctly", () => {
  const { asFragment } = render(
    <Tag>
      <TagLabel>B</TagLabel>
      <TagCloseButton />
    </Tag>,
  )
  expect(asFragment()).toMatchSnapshot()
})

test("Tag with left icon renders correctly", () => {
  const { asFragment } = render(
    <Tag>
      <TagLeftIcon />
      <TagLabel>C</TagLabel>
      <TagCloseButton />
    </Tag>,
  )
  expect(asFragment()).toMatchSnapshot()
})
