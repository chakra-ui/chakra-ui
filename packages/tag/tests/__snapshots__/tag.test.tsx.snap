// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag (only) renders correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: top;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: 100%;
  font-weight: 500;
  line-height: 1.2;
  outline: 0;
  min-height: 2rem;
  min-width: 2rem;
  font-size: 1rem;
  border-radius: 0.25rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  background: #EDF2F7;
  color: #1A202C;
}

.emotion-0:focus,
.emotion-0[data-focus] {
  box-shadow: 0 0 0 3px rgba(66,153,225,0.6);
}

<span
    class="emotion-0"
  >
    A
  </span>
</DocumentFragment>
`;

exports[`Tag with close button renders correctly 1`] = `
<DocumentFragment>
  .emotion-3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: top;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: 100%;
  font-weight: 500;
  line-height: 1.2;
  outline: 0;
  min-height: 2rem;
  min-width: 2rem;
  font-size: 1rem;
  border-radius: 0.25rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  background: #EDF2F7;
  color: #1A202C;
}

.emotion-3:focus,
.emotion-3[data-focus] {
  box-shadow: 0 0 0 3px rgba(66,153,225,0.6);
}

.emotion-0 {
  line-height: 1.2;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  outline: 0;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  font-size: 1em;
  width: 1.25rem;
  height: 1.25rem;
  border-radius: 0.125rem;
  margin-left: 0.375rem;
  opacity: 0.5;
}

.emotion-2[disabled],
.emotion-2[disabled]:focus,
.emotion-2[disabled]:hover,
.emotion-2[aria-disabled=true],
.emotion-2[aria-disabled=true]:focus,
.emotion-2[aria-disabled=true]:hover,
.emotion-2[data-disabled],
.emotion-2[data-disabled]:focus,
.emotion-2[data-disabled]:hover {
  opacity: 0.4;
}

.emotion-2:focus,
.emotion-2[data-focus] {
  box-shadow: 0 0 0 3px rgba(66,153,225,0.6);
  background: rgba(0,0,0,0.14);
}

.emotion-2:hover,
.emotion-2[data-hover] {
  opacity: 0.8;
}

.emotion-2:active,
.emotion-2[data-active] {
  opacity: 1;
}

.emotion-1 {
  vertical-align: inherit;
  display: inline-block;
  line-height: 1em;
  color: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 100%;
  height: 100%;
}

<span
    class="emotion-3"
  >
    <span
      class="emotion-0"
    >
      B
    </span>
    <button
      class="emotion-2"
      type="button"
    >
      <svg
        class="chakra-icon emotion-1"
        focusable="false"
        role="presentation"
        viewBox="0 0 512 512"
      >
        <path
          d="M289.94 256l95-95A24 24 0 00351 127l-95 95-95-95a24 24 0 00-34 34l95 95-95 95a24 24 0 1034 34l95-95 95 95a24 24 0 0034-34z"
          fill="currentColor"
        />
      </svg>
    </button>
  </span>
</DocumentFragment>
`;

exports[`Tag with left icon renders correctly 1`] = `
<DocumentFragment>
  .emotion-4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: top;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: 100%;
  font-weight: 500;
  line-height: 1.2;
  outline: 0;
  min-height: 2rem;
  min-width: 2rem;
  font-size: 1rem;
  border-radius: 0.25rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  background: #EDF2F7;
  color: #1A202C;
}

.emotion-4:focus,
.emotion-4[data-focus] {
  box-shadow: 0 0 0 3px rgba(66,153,225,0.6);
}

.emotion-1 {
  line-height: 1.2;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  outline: 0;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  font-size: 1em;
  width: 1.25rem;
  height: 1.25rem;
  border-radius: 0.125rem;
  margin-left: 0.375rem;
  opacity: 0.5;
}

.emotion-3[disabled],
.emotion-3[disabled]:focus,
.emotion-3[disabled]:hover,
.emotion-3[aria-disabled=true],
.emotion-3[aria-disabled=true]:focus,
.emotion-3[aria-disabled=true]:hover,
.emotion-3[data-disabled],
.emotion-3[data-disabled]:focus,
.emotion-3[data-disabled]:hover {
  opacity: 0.4;
}

.emotion-3:focus,
.emotion-3[data-focus] {
  box-shadow: 0 0 0 3px rgba(66,153,225,0.6);
  background: rgba(0,0,0,0.14);
}

.emotion-3:hover,
.emotion-3[data-hover] {
  opacity: 0.8;
}

.emotion-3:active,
.emotion-3[data-active] {
  opacity: 1;
}

.emotion-2 {
  vertical-align: inherit;
  display: inline-block;
  line-height: 1em;
  color: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 100%;
  height: 100%;
}

.emotion-0 {
  vertical-align: top;
  display: inline-block;
  line-height: 1em;
  color: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 1em;
  height: 1em;
  margin-right: 0.5rem;
}

<span
    class="emotion-4"
  >
    <svg
      class="chakra-icon emotion-0"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <g
        stroke="currentColor"
        stroke-width="1.5"
      >
        <path
          d="M9,9a3,3,0,1,1,4,2.829,1.5,1.5,0,0,0-1,1.415V14.25"
          fill="none"
          stroke-linecap="round"
        />
        <path
          d="M12,17.25a.375.375,0,1,0,.375.375A.375.375,0,0,0,12,17.25h0"
          fill="currentColor"
          stroke-linecap="round"
        />
        <circle
          cx="12"
          cy="12"
          fill="none"
          r="11.25"
          stroke-miterlimit="10"
        />
      </g>
    </svg>
    <span
      class="emotion-1"
    >
      C
    </span>
    <button
      class="emotion-3"
      type="button"
    >
      <svg
        class="chakra-icon emotion-2"
        focusable="false"
        role="presentation"
        viewBox="0 0 512 512"
      >
        <path
          d="M289.94 256l95-95A24 24 0 00351 127l-95 95-95-95a24 24 0 00-34 34l95 95-95 95a24 24 0 1034 34l95-95 95 95a24 24 0 0034-34z"
          fill="currentColor"
        />
      </svg>
    </button>
  </span>
</DocumentFragment>
`;
